Main.cpp

```diff
@@ -182,52 +233,50 @@ int main() {
 // @return temp Temperature in degrees C, resolution of .25
-uint8_t adc_to_temp(uint8_t* rxbuf) {
-  static uint16_t raw{};
-  static int norm{};
+// TODO: Add in per-channel moving average
+uint8_t adcToTemp(uint8_t* rxbuf) {
   // voltage at max supported temperature (63.75C)
-  static constexpr float kMaxTempVoltage = 1.4875;
-  static constexpr float kMinTempVoltage = 2.17;
-  static constexpr float kVRef = 3.316;
+  static constexpr float kVref = 3.315;
   static constexpr float kAdcMax = 1023.0;
-  // true resolution of ~ 211.5 values across 0-63.75C
-  static constexpr float kMaxTempValue =
-      (kMaxTempVoltage / kVRef) * kAdcMax;  // ~ 461.125
-  static constexpr float kMinTempValue =
-      (kMinTempVoltage / kVRef) * kAdcMax;  // ~ 672.7
-  static constexpr float kMaxIntegerValue =
-      kMinTempValue - kMaxTempValue;  // ~ 211.575
-  // unpack raw reading to continuous 10 bits
-  // A) ((lower 7 of byte 0) << 3) | (lower 3 of byte 1)
-  // B) (upper 7 bits of the reading, in the upper 7 position) |
-  //      (lower 3 bits of the second byte, in the lower 3 position)
-  //
-  raw = ((rxbuf[0] & 0x7f) << 3) | (rxbuf[1] & 0x7);
-  // normalize to 0=63.75C to 255=0C
-  // (decode to degree C by dividing by 4...bit shifting right by 2)
-  norm = ((raw - kMaxTempValue) / kMaxIntegerValue) * 255;
-  // top-out temp at 63.75C and bottom-out at 0C
-  if (norm < 0) {
-    norm = 0;
-  } else if (norm > 255) {
-    norm = 255;
-  }
-  // normalize to 0=0C 255=63.75C and return
-  return (255 - norm);
+  static constexpr uint8_t kMaxTemp = 255;  // max of 63.75C
+
+  // Pack bits from SPI bytes into single 10b ADC reading:
+  // (upper 7 bits of the first byte, in the upper 7 position) |
+  //    (upper 3 bits of the second byte, in the lower 3 position)
+  uint16_t raw = ((rxbuf[0] & 0x7f) << 3) | ((rxbuf[1] & 0xe0) >> 5);
+
+  // convert raw reading to voltage based on ADC resolution and Vref
+  float voltage = (raw / kAdcMax) * kVref;
+
+  // Convert fp voltage to integer (with 0.01V resolution), then normalize to
+  // 0=148 (1.48V) and 69=217 (2.17V)
+  // Voltages below 0C (> 2.17V) are unlikely and irrelevant to fault detection.
+  // Voltages above 63.75C (< 1.48V) are implausible due to system fault @ 55C.
+  int lookupIndex = voltage * 100 - 148;
+
+  // clamp index between 0 and 69
+  lookupIndex = std::clamp<uint16_t>(lookupIndex, 0, g_kMaxVoltageIndex);
+
+  // fetch linearly interpolable range (e.g. 0-5 degrees C)
+  auto& interpolationSegment = g_voltageToTemp[lookupIndex];
+
+  // TODO: change to uint8_t (need to test effect on temp resolution)
+  // compute temperature offset from upper bound of range
+  float interpolatedBaseOffset = 5 *
+                                 (voltage * 100 - interpolationSegment.start) /
+                                 interpolationSegment.length;
+
+  // compute fp temperature from base and base offset
+  uint16_t temp = 4 * (interpolationSegment.baseTemp - interpolatedBaseOffset);
```

Comment:
That comment doesn't belong there; I'll change it. The 4 is to scale the 0-63.75 range to 0-255 (while maintain .25C resolution w.r.t the fp equivalent), along with facilitating production of sufficient fairy dust